framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceImplementation.java

```diff
@@ -743,13 +767,93 @@ public void addMethodDeclarationAnnotation(ExecutableElement methodElt, Annotati
       return;
     }
 
+    // Special-case handling for purity annotations.
+    AnnotationMirror annoToAdd;
+    if (!isPurityAnno(anno)) {
+      annoToAdd = anno;
+    } else {
+      // It's a purity annotation. Do a "least upper bound" between the current purity annotation
+      // inferred for the method and anno. This is necessary to avoid WPI inferring incompatible
+      // purity annotations on methods that override methods from their superclass. TODO: this would
+      // be unnecessary if purity implemented as a type system.
+      AnnotationMirror currentPurityAnno = getPurityAnnotation(methodElt);
+      if (currentPurityAnno == null) {
+        annoToAdd = anno;
+      } else {
+        // Clear the current purity annotation, because at this point a new one is definitely
+        // going to be inferred.
+        storage.removeMethodDeclarationAnnotation(methodElt, currentPurityAnno);
+
+        // TODO: is this the best way to do this? Would it be easier to just write a subtype
```

Comment:
I suggest abstracting this lub into a separate routine, which will make this one shorter and easier to understand.